### PR TITLE
Vecops consistent cell default add more checking on location

### PR DESCRIPTION
**Pull request by d7919**
Provides bugfixes to ensure the vector result has location set.

Also ensure copy/assignment of vectors updates the location of the result.

### CODE REVIEW

File: src/field/vecops.cxx [hunk 1]
```diff
@@ -82,35 +81,33 @@ const Vector3D Grad(const Field3D &f, CELL_LOC outloc) {
...
   result.x = DDX(f, outloc_x) -
              metric_x->g_12 * DDY(f, outloc_x) / SQ(metric_x->J * metric_x->Bxy);
   result.y = 0.0;
   result.z = DDZ(f, outloc_z) -
              metric_z->g_23 * DDY(f, outloc_z) / SQ(metric_z->J * metric_z->Bxy);
 
+  CELL_LOC outloc = (outloc_x == outloc_y && outloc_x == outloc_z)
+                        ? result.x.getLocation()
+                        : CELL_VSHIFT;
+  result.setLocation(outloc);
```
Comment (ZedThree) [inline src/field/vecops.cxx]: Do we also need to make sure the individual components are now all at the same/consistent location as well? (or `result.y` at least?)

Comment (d7919) [reply src/field/vecops.cxx]: Nope, `result.setLocation` sets the location on each of the components.

File: src/field/vecops.cxx [hunk 2]
```diff
@@ -175,14 +180,18 @@ const Field3D Div(const Vector3D &v, CELL_LOC outloc) {
...
   if (outloc == CELL_DEFAULT) {
     outloc = v.getLocation();
   }
 
+  ASSERT1(outloc != CELL_VSHIFT);
+
+  // The following probably belongs in the FDD? routines rather than here
+  // as we don't directly interact v and f. Note also no equivalent in Field3D
+  // version of this routine.
+  ASSERT1(outloc == f.getLocation());
```
Comment (ZedThree) [inline src/field/vecops.cxx]: Could you say why the Field3D version doesn't have this?

Comment (d7919) [reply src/field/vecops.cxx]: I don't know why -- it's just an observation to motivate why we might want to remove/move this!

Comment (johnomotani) [reply src/field/vecops.cxx]: This is checked by indexFDD*, so we can remove it from here :+1:

File: src/field/vecops.cxx [hunk 3]
```diff
@@ -195,39 +204,45 @@ const Field2D Div(const Vector2D &v, const Field2D &f, CELL_LOC outloc) {
   result += FDDZ(metric->J * vcn.z, f, outloc);
   result /= metric->J;
 
+  result.setLocation(outloc);
```
Comment (johnomotani) [inline src/field/vecops.cxx]: This is not necessary, result.location is set already by FDDX. In the Field*D difops this is just skipped, but we could put an `ASSERT1(result.getLocation()==outloc)` if you want to double-check?

File: src/field/vecops.cxx [hunk 4]
```diff
@@ -195,39 +204,45 @@ const Field2D Div(const Vector2D &v, const Field2D &f, CELL_LOC outloc) {
...
   Vector3D vcn = v;
   vcn.toContravariant();
 
+  Field3D result(localmesh);
   result = FDDX(metric->J * vcn.x, f, outloc, method);
   result += FDDY(metric->J * vcn.y, f, outloc, method);
   result += FDDZ(metric->J * vcn.z, f, outloc, method);
   result /= metric->J;
 
+  result.setLocation(outloc);
```
Comment (johnomotani) [inline src/field/vecops.cxx]: As above, this is not necessary.

File: src/field/vecops.cxx (hunk 3 above)
Comment (d7919) [reply src/field/vecops.cxx]: Ah yes the return is a Field not a vector -- I was busy adding these for the routines which return vectors and must have got carried away!

Comment (d7919) [reply src/field/vecops.cxx]: Actually (at least in this branch) FDDX (and indexFDDX) on field2D doesn't set the location as far as I can tell.

Comment (d7919) [reply src/field/vecops.cxx]: @ZedThree, @johnomotani  should I fix this missing `setLocation` here or in a separate PR?

Comment (d7919) [reply src/field/vecops.cxx]: Should note the 2D/3D DDX versions shouldn't need the location setting in vecops either as these are both set in either `indexDDX` and/or `applyXDiff`. It's just `indexFDDX(Field2D)` that doesn't set a location currently.

**Comment by d7919** [comment]
Just to note that this requires #1306 to ensure location of `Field2D` returned by the `Div` that uses flux methods (e.g. `indexFDDX`) is set.

File: src/field/vecops.cxx [hunk 5]
```diff
@@ -35,43 +35,42 @@
...
-  
-  if (outloc == CELL_DEFAULT) {
-    outloc = f.getLocation();
-  }
+
+  Vector2D result(f.getMesh());
 
   result.x = DDX(f, outloc);
   result.y = DDY(f, outloc);
   result.z = DDZ(f, outloc);
```
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: Vector2D can indeed have VSHIFT but there is no handling of it anywhere currently.

We could just use `result.setLocation(result.x.getLocation());` to set the location at the end in the meantime. Currently VSHIFT isn't supported for this routine (it should throw if you try) -- changing this may be regarded as a bug fix or it may be a feature!

[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: I agree about trying to get rid of the three argument versions.

File: src/field/vecops.cxx [hunk 6]
```diff
@@ -123,13 +120,15 @@ const Vector3D Grad_perp(const Field3D &f, CELL_LOC outloc_x, CELL_LOC UNUSED(ou
 const Field2D Div(const Vector2D &v, CELL_LOC outloc) {
   TRACE("Div( Vector2D )");
 
-  Mesh *localmesh = v.x.getMesh();
-  Field2D result(localmesh);
-
   if (outloc == CELL_DEFAULT) {
     outloc = v.getLocation();
```
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: I'm not sure, I think if the vector is at VSHIFT then I'd expect this to fail

File: src/field/vecops.cxx [hunk 7]
```diff
@@ -239,12 +241,13 @@ const Vector2D Curl(const Vector2D &v, CELL_LOC outloc) {
 
   TRACE("Curl( Vector2D )");
 
-  Mesh *localmesh = v.x.getMesh();
-
   if (outloc == CELL_DEFAULT) {
     outloc = v.getLocation();
```
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: Not sure if this comment is to contrast the earlier ones, but just in case you didn't spot it the subsequent line asserts that the location isn't vshift, so will indeed fail.

File: src/field/vecops.cxx [hunk 8]
```diff
@@ -374,9 +374,15 @@ const Field3D V_dot_Grad(const Vector3D &v, const Field3D &f) {
...
 }
 
-const Vector2D V_dot_Grad(const Vector2D &v, const Vector2D &a, const CELL_LOC outloc) {
+const Vector2D V_dot_Grad(const Vector2D &v, const Vector2D &a, CELL_LOC outloc) {
   TRACE("V_dot_Grad( Vector2D , Vector2D )");
 
+  ASSERT1(outloc != CELL_VSHIFT);
+  if (outloc == CELL_DEFAULT) {
+    ASSERT1(outloc == v.getLocation() && outloc == a.getLocation());
+    outloc = v.getLocation();
```
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: Hmm, that does look strange I think it should perhaps be ?

~~~~
if (outloc == CELL_DEFAULT) {
    ASSERT1(v.getLocation() == a.getLocation());
    outloc = v.getLocation();
}
~~~~

[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: and we probably need to assert that v.getLocation isn't VSHIFT

[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: (possibly just move the outloc assert after the if block?)

File: src/field/vecops.cxx [hunk 9]
```diff
@@ -35,43 +35,42 @@
...
 
+  result.setLocation(outloc);
+
   result.covariant = true;
 
   return result;
 }
 
 const Vector3D Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
                     CELL_LOC outloc_z) {
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: The three-arg versions can be deprecated for this release

File: src/field/vecops.cxx [hunk 10]
```diff
@@ -82,35 +81,33 @@ const Vector3D Grad(const Field3D &f, CELL_LOC outloc) {
...
+          (outloc_x == CELL_XLOW && outloc_y == CELL_YLOW &&
+           outloc_z == CELL_ZLOW)); // CELL_VSHIFT
 
-  Coordinates *metric_x = mesh->coordinates(outloc_x);
-  Coordinates *metric_z = mesh->coordinates(outloc_z);
+  Coordinates *metric_x = f.getCoordinates(outloc_x);
+  Coordinates *metric_z = f.getCoordinates(outloc_z);
 
   result.x = DDX(f, outloc_x) -
              metric_x->g_12 * DDY(f, outloc_x) / SQ(metric_x->J * metric_x->Bxy);
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: That's a pain. Maybe a check with a nice exception message to say that setting the location is currently unsupported?

**Comment by d7919** [comment]
I've just pushed something that demos adding CELL_VSHIFT support to Grad(Field2D) and deprecates the three arg version of Grad(Field3D).

I'll try to deprecate the other three arg routines, but let me know if you have any comments on these most recent changes.

File: src/field/vecops.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: I think failing is the correct option here. The documentation might need to be updated to say that it's not supported.

File: src/field/vecops.cxx (hunk 10 above)
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: Have just pushed a commit that tries to address this.

[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: Actually, this is more than a pain, it means `CELL_VSHIFT` is not terribly useful. Unfortunately, numerically, it is a useful staggering option. Sounds like we will need to put some future work into trying to support interpolating to orthogonal directions in order to better support `CELL_VSHIFT`

File: src/field/vecops.cxx [hunk 11]
```diff
@@ -37,76 +37,97 @@
...
+  }
 
   result.covariant = true;
 
   return result;
 }
 
-const Vector3D Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
-                    CELL_LOC outloc_z) {
+const Vector3D DEPRECATED(Grad(const Field3D &f, CELL_LOC outloc_x, CELL_LOC outloc_y,
```
Comment (ZedThree) [inline src/field/vecops.cxx]: This should go on the declaration in the header rather than the implementation.

Also, the single-arg version needs to go in the header. Note that I think you'll need to remove the default values for the three-arg version and add one for the single-arg.

Comment (d7919) [reply src/field/vecops.cxx]: Ah, forgot the header.

Should I remove deprecated from the source file and just put it on header?

Comment (d7919) [reply src/field/vecops.cxx]: Also doesn't removing the default values break backwards compatibility?

File: src/field/vecops.cxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: What combination of three locations are supported? 

- `v` and `a` at the same location, and `outloc == CELL_DEFAULT` -- this should be fine
- everything at `CELL_CENTRE` -- this should be fine
- everything at other locations? -- presumably fine?
- `v` and `a` at `CELL_CENTRE`, `outloc == CELL_?LOW`? vice-versa? -- don't know
- `v` at `CELL_CENTRE`, `a` staggered (and vice-versa), `outloc == CELL_DEFAULT`? -- don't know

If just first two, bring the assert above the `if`. Not sure about other cases yet.

[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: 0669941 should address this. Essentially v and a have to be at the same location and outloc has to also refer to this location (or CELL_DEFAULT).

[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: Just to clarify, just the first three combinations above are supported?

There's a fair bit of information about supported locations coming out of the woodwork in this PR -- can some of it go in the doxygen comments?

[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: v and a at the same location, and `outloc == CELL_DEFAULT` or `outloc==v.getLocation()`.

**Comment by d7919** [comment]
Turns out the `outloc` argument for `V_dot_Grad` does not appear in the header for vecops.

File: src/field/vecops.cxx (hunk 8 above)
[the remark this answers is not in the document]
Comment (d7919) [reply src/field/vecops.cxx]: I've just added some brief doxygen comments.

**Comment by ZedThree** [comment]
So it doesn't! It's currently not possible to call `V_dot_Grad(vector, vector)`! Yes, remove it from the implementation please!

Looks like it came in in 6fd914d, #1200

**Comment by d7919** [comment]
> So it doesn't! It's currently not possible to call `V_dot_Grad(vector, vector)`! Yes, remove it from the implementation please!
> 
> Looks like it came in in [6fd914d](https://github.com/boutproject/BOUT-dev/commit/6fd914dcae1e74ad9353f97f5a66730b4b9110aa), #1200

So I actually just added it to the header -- have to dash now but feel free to remove this as you see fit!

**Comment by ZedThree** [comment]
If I understand correctly, it basically doesn't make sense for `V_dot_Grad` to take a location, for any of the overloads? In which case, let's remove the location completely

This also highlights that we don't have any tests for `V_dot_Grad`, as it's not currently possible to use it!

File: src/field/vecops.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/field/vecops.cxx]: Second-guessing myself a bit here, but I think failing for `CELL_DEFAULT` is ok. Passing `CELL_CENTRE` explicitly should work at any rate.

**Comment by ZedThree** [comment]
Some big changes here. @johnomotani @d7919 are you both happy with the current state? I'm sure we've missed some edge cases, but vecops is lacking tests -- needs to be remedied for 4.3

File: src/field/vecops.cxx [hunk 12]
```diff
@@ -396,250 +396,216 @@ const Field3D V_dot_Grad(const Vector3D &v, const Field3D &f) {
...
-  // Note the following assert means the outloc argument is pointless
-  ASSERT1(v.getLocation() == outloc);
 
-  Mesh *localmesh = v.x.getMesh();
-  Vector2D result(localmesh);
+  Vector2D result{v.x.getMesh()};
 
-  Coordinates *metric = localmesh->coordinates(outloc);
+  auto metric_x = v.x.getCoordinates();
+  auto metric_y = v.y.getCoordinates();
```
Comment (d7919) [inline src/field/vecops.cxx]: These all have to be at exactly the same location so we only really need one of the three.

File: src/field/vecops.cxx [hunk 13]
```diff
@@ -396,250 +396,216 @@ const Field3D V_dot_Grad(const Vector3D &v, const Field3D &f) {
...
+
+    result.z = VDDX(vcn.x, a.z) + VDDY(vcn.y, a.z) + VDDZ(vcn.z, a.z);
+    result.z += vcn.x * (metric_x->G3_11 * a.x + metric_y->G3_12 * a.y + metric_z->G3_13 * a.z);
+    result.z += vcn.y * (metric_x->G3_12 * a.x + metric_y->G3_22 * a.y + metric_z->G3_23 * a.z);
+    result.z += vcn.z * (metric_x->G3_13 * a.x + metric_y->G3_23 * a.y + metric_z->G3_33 * a.z);
 
     result.covariant = false;
   }
 
-  result.setLocation(outloc);
```
Comment (d7919) [inline src/field/vecops.cxx]: I think we still need to set the location of the resultant vector otherwise it's exactly `CELL_CENTRE` but inputs might both be at `CELL_ZLOW` for example.

File: src/field/vecops.cxx (hunk 12 above)
Comment (ZedThree) [reply src/field/vecops.cxx]: Ah, I didn't catch that it doesn't support `VSHIFT` for the input locations

**Comment by d7919** [comment]
I'm going to merge this shortly into the branch for #1293 and then we can review that PR in its entirety.
